src/nvenc/nvenc_base.cpp

```diff
@@ -0,0 +1,323 @@
+#include "nvenc_base.h"
+
+namespace {
+  GUID
+  quality_preset_guid_from_number(unsigned number) {
+    if (number > 7) number = 7;
+
+    switch (number) {
+      case 1:
+      default:
+        return NV_ENC_PRESET_P1_GUID;
+
+      case 2:
+        return NV_ENC_PRESET_P2_GUID;
+
+      case 3:
+        return NV_ENC_PRESET_P3_GUID;
+
+      case 4:
+        return NV_ENC_PRESET_P4_GUID;
+
+      case 5:
+        return NV_ENC_PRESET_P5_GUID;
+
+      case 6:
+        return NV_ENC_PRESET_P6_GUID;
+
+      case 7:
+        return NV_ENC_PRESET_P7_GUID;
+    }
+  };
+
+  auto
+  quality_preset_string_from_guid(const GUID &guid) {
+    if (guid == NV_ENC_PRESET_P1_GUID) {
+      return "P1";
+    }
+    if (guid == NV_ENC_PRESET_P2_GUID) {
+      return "P2";
+    }
+    if (guid == NV_ENC_PRESET_P3_GUID) {
+      return "P3";
+    }
+    if (guid == NV_ENC_PRESET_P4_GUID) {
+      return "P4";
+    }
+    if (guid == NV_ENC_PRESET_P5_GUID) {
+      return "P5";
+    }
+    if (guid == NV_ENC_PRESET_P6_GUID) {
+      return "P6";
+    }
+    if (guid == NV_ENC_PRESET_P7_GUID) {
+      return "P7";
+    }
+    return "Unknown";
+  }
+
+}  // namespace
+
+namespace nvenc {
+
+  nvenc_base::nvenc_base(NV_ENC_DEVICE_TYPE device_type, void *device, uint32_t max_width, uint32_t max_height, NV_ENC_BUFFER_FORMAT buffer_format):
+      device_type(device_type),
+      device(device),
+      max_width(max_width),
+      max_height(max_height),
+      buffer_format(buffer_format) {
+  }
+
+  nvenc_base::~nvenc_base() {
+    // Use destroy_base_resources() instead
+  }
+
+  bool
+  nvenc_base::create_encoder(const nvenc_config &config, const video::config_t &client_config, const nvenc_colorspace &colorspace) {
+    if (encoder) return false;
+    if (!nvenc && !init_library()) return false;
+
+    NV_ENC_OPEN_ENCODE_SESSION_EX_PARAMS session_params = { NV_ENC_OPEN_ENCODE_SESSION_EX_PARAMS_VER };
+    session_params.device = device;
+    session_params.deviceType = device_type;
+    session_params.apiVersion = NVENCAPI_VERSION;
+    if (nvenc->nvEncOpenEncodeSessionEx(&session_params, &encoder) != NV_ENC_SUCCESS) {
+      BOOST_LOG(error) << "NvEncOpenEncodeSessionEx failed";
+      return false;
+    }
+
+    width = client_config.width;
+    height = client_config.height;
+
+    NV_ENC_INITIALIZE_PARAMS init_params = { NV_ENC_INITIALIZE_PARAMS_VER };
+    init_params.maxEncodeWidth = max_width;
+    init_params.maxEncodeHeight = max_height;
+    init_params.presetGUID = quality_preset_guid_from_number(config.quality_preset);
+    init_params.tuningInfo = NV_ENC_TUNING_INFO_ULTRA_LOW_LATENCY;

```

Comment:
Can you plumb in this `tuningInfo` value from the config? We expose this option to users as `config::video.nv.nv_tune`.

Reply:
Close to final concept image with the new options page https://github.com/LizardByte/Sunshine/pull/1427#issuecomment-1640116140